Lib/idlelib/idle_test/test_debugger.py

```diff
@@ -23,7 +34,456 @@ def test_init(self):
+
+        # Create a test code object to simulate a debug session.
+        code_obj = compile(TEST_CODE,
+                           'test_user_line_basic_frame.py',
+                           mode='exec')
+
+        # Create 2 test frames for lines 1 and 2 of the test code.
+        test_frame1 = MockFrameType(code_obj, 1)
+
+        test_frame2 = MockFrameType(code_obj, 2)
+        test_frame2.f_back = test_frame1
+
+        gui = mock.Mock()
+        gui.interaction = mock.Mock()
+
+        idb = debugger.Idb(gui)
+        idb.user_line(test_frame2)
+
+        self.assertFalse(idb.in_rpc_code(test_frame2))
+        gui.interaction.assert_called_once()
+        gui.interaction.assert_called_with('test_user_line_basic_frame.py:2: <module>()', test_frame2)
+
+    def test_user_exception(self):
+        """Test that .user_exception() creates a string message for a frame."""
+
+        # Create a test code object to simulate a debug session.
+        code_obj = compile(TEST_CODE,
+                           'test_user_exception.py',
+                           mode='exec')
+
+        # Create 1 test frame
+        test_frame1 = MockFrameType(code_obj, 1)
+
+        # Example from sys.exc_info()
+        test_exc_info = (type(ValueError), ValueError(), None)
+
+        gui = mock.Mock()
+        gui.interaction = mock.Mock()
+
+        idb = debugger.Idb(gui)
+        idb.user_exception(test_frame1, test_exc_info)
+
+        self.assertFalse(idb.in_rpc_code(test_frame1))
+        gui.interaction.assert_called_once()
+        gui.interaction.assert_called_with('test_user_exception.py:1: <module>()', test_frame1, test_exc_info)
+
+    def test_in_rpc_code_rpc_py(self):
+        """Test that .in_rpc_code detects position of rpc.py."""
+
+        # Create a test code object to simulate a debug session.
+        code_obj = compile(TEST_CODE,
+                           'rpc.py',
+                           mode='exec')
+
+        # Create 1 test frame
+        test_frame = MockFrameType(code_obj, 1)
+
+        gui = mock.Mock()
+        gui.interaction = mock.Mock()
+
+        idb = debugger.Idb(gui)
+
+        self.assertTrue(idb.in_rpc_code(test_frame))
+
+    def test_in_rpc_code_debugger_star_dot_py(self):
+        """Test that .in_rpc_code detects position of idlelib/debugger*.py."""
+
+        # Create a test code object to simulate a debug session.
+        for filename in ('idlelib/debugger.py', 'idlelib/debugger_r.py'):
+
+            code_obj = compile(TEST_CODE,
+                               filename,
+                               mode='exec')
+
+            # Create 2 test frames
+            test_frame = MockFrameType(code_obj, 1)
+
+            test_frame2 = MockFrameType(code_obj, 2)
+            test_frame2.f_back = test_frame
+
+            gui = mock.Mock()
+            gui.interaction = mock.Mock()
+
+            idb = debugger.Idb(gui)
+
+            self.assertFalse(idb.in_rpc_code(test_frame2))
+
+
+def make_pyshell_mock():
```

Comment:
There's already a mock Editor class in `mock_idle.py`, so it may be helpful to create a shell mock there.  That way, it can be used by other modules if they need a mock shell for testing.

Reply:
Fixed